// ngx-webstorage/esm5/stubs/storageStrategy.stub.js
import { of, Subject } from 'rxjs';
export var StorageStrategyStubName = 'stub_strategy';
var StorageStrategyStub = /** @class */ (function () {
    function StorageStrategyStub(name) {
        this.keyChanges = new Subject();
        this.store = {};
        this._available = true;
        this.name = name || StorageStrategyStubName;
    }
    Object.defineProperty(StorageStrategyStub.prototype, "isAvailable", {
        get: function () {
            return this._available;
        },
        enumerable: true,
        configurable: true
    });
    StorageStrategyStub.prototype.get = function (key) {
        return of(this.store[key]);
    };
    StorageStrategyStub.prototype.set = function (key, value) {
        this.store[key] = value;
        this.keyChanges.next(key);
        return of(value);
    };
    StorageStrategyStub.prototype.del = function (key) {
        delete this.store[key];
        this.keyChanges.next(key);
        return of(null);
    };
    StorageStrategyStub.prototype.clear = function () {
        this.store = {};
        this.keyChanges.next(null);
        return of(null);
    };
    return StorageStrategyStub;
}());
export { StorageStrategyStub };
