// ngx-webstorage/esm5/lib/core/templates/asyncStorage.js
import { StorageKeyManager } from '../../helpers/storageKeyManager';
import { distinctUntilChanged, filter, map, shareReplay, switchMap } from 'rxjs/operators';
var AsyncStorage = /** @class */ (function () {
    function AsyncStorage(strategy) {
        this.strategy = strategy;
    }
    AsyncStorage.prototype.retrieve = function (key) {
        return this.strategy.get(StorageKeyManager.normalize(key)).pipe(map(function (value) { return typeof value === 'undefined' ? null : value; }));
    };
    AsyncStorage.prototype.store = function (key, value) {
        return this.strategy.set(StorageKeyManager.normalize(key), value);
    };
    AsyncStorage.prototype.clear = function (key) {
        return key !== undefined ? this.strategy.del(StorageKeyManager.normalize(key)) : this.strategy.clear();
    };
    AsyncStorage.prototype.getStrategyName = function () { return this.strategy.name; };
    AsyncStorage.prototype.observe = function (key) {
        var _this = this;
        key = StorageKeyManager.normalize(key);
        return this.strategy.keyChanges.pipe(filter(function (changed) { return changed === null || changed === key; }), switchMap(function () { return _this.strategy.get(key); }), distinctUntilChanged(), shareReplay());
    };
    return AsyncStorage;
}());
export { AsyncStorage };
