// ngx-webstorage/esm5/lib/strategies/baseSyncStorage.js
import { of, Subject } from 'rxjs';
import { CompatHelper } from '../helpers/compat';
var BaseSyncStorageStrategy = /** @class */ (function () {
    function BaseSyncStorageStrategy(storage, cache) {
        this.storage = storage;
        this.cache = cache;
        this.keyChanges = new Subject();
    }
    Object.defineProperty(BaseSyncStorageStrategy.prototype, "isAvailable", {
        get: function () {
            if (this._isAvailable === undefined)
                this._isAvailable = CompatHelper.isStorageAvailable(this.storage);
            return this._isAvailable;
        },
        enumerable: true,
        configurable: true
    });
    BaseSyncStorageStrategy.prototype.get = function (key) {
        var data = this.cache.get(this.name, key);
        if (data !== undefined)
            return of(data);
        try {
            var item = this.storage.getItem(key);
            if (item !== null) {
                data = JSON.parse(item);
                this.cache.set(this.name, key, data);
            }
        }
        catch (err) {
            console.warn(err);
        }
        return of(data);
    };
    BaseSyncStorageStrategy.prototype.set = function (key, value) {
        var data = JSON.stringify(value);
        this.storage.setItem(key, data);
        this.cache.set(this.name, key, value);
        this.keyChanges.next(key);
        return of(value);
    };
    BaseSyncStorageStrategy.prototype.del = function (key) {
        this.storage.removeItem(key);
        this.cache.del(this.name, key);
        this.keyChanges.next(key);
        return of(null);
    };
    BaseSyncStorageStrategy.prototype.clear = function () {
        this.storage.clear();
        this.cache.clear(this.name);
        this.keyChanges.next(null);
        return of(null);
    };
    return BaseSyncStorageStrategy;
}());
export { BaseSyncStorageStrategy };
